### PR TITLE
:building_construction: Created the Jira Assets module

### PR DESCRIPTION
**Pull request by ctreminiom**
1. Created the service interfaces for the Assets icons,AQL, and object's.

### CODE REVIEW

File: service/assets/icon.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+	// Get loads a single asset icon by id.
+	//
+	// GET /jsm/assets/workspace/{workspaceId}/v1/icon/{id}
+	Get(ctx context.Context, workspaceID, iconID string) (*models.IconAssetScheme, *models.ResponseScheme, error)
+
+	// Global returns all global icons i.e. icons not associated with a particular object schema.
+	//
+	// GET /jsm/assets/workspace/{workspaceId}/v1/icon/global
+	//
+	Global(ctx context.Context, workspaceID string) (*[]models.IconAssetScheme, *models.ResponseScheme, error)
```
Comment (Fank) [inline service/assets/icon.go:21]: You are using a pointer before the array, compared to the code in service/assets/object.go it looks like it should be
```suggestion
	Global(ctx context.Context, workspaceID string) ([]*models.IconAssetScheme, *models.ResponseScheme, error)
```

**Review by Fank — CHANGES_REQUESTED** [review]

**Comment by ctreminiom** [comment]
@Fank the first version of the Jira Assets implementation is now operative!, I tested and document each endpoint and it works as expected. 

But, as I'm not a Jira Asset expert, could you please test the methods and provide some feedback about the module?

The documentation is hosted at https://docs.go-atlassian.io/jira-assets/overview

_Pura Vida_

**Comment by Fank** [comment]
Thanks @ctreminiom

I started replacing everything in our code so far, but ran into some issues i opened in seperate issues here.
I will just replace assets and give you feedback about it soon as i can, expect ~2 weeks.
Keep the PR open, i will comment about everything i find.

File: service/assets/aql.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+// AQLAssetConnector represents the assets AQL search endpoints.
+// Use it to execute AQL searches.
+type AQLAssetConnector interface {
+
+	// Filter retrieves a list of objects based on an AQL. Note that the preferred endpoint is /aql.
+	//
+	// POST /jsm/assets/workspace/{workspaceId}/v1/object/navlist/aql
+	//
+	// https://docs.go-atlassian.io/jira-assets/aql#filter-objects
```
Comment (Fank) [inline service/assets/aql.go:16]:
```suggestion
	// Filter retrieves a list of objects based on an AQL. Note that the preferred endpoint is /aql.
	//
	// POST /jsm/assets/workspace/{workspaceId}/v1/object/navlist/aql
	//
	// https://docs.go-atlassian.io/jira-assets/aql#filter-objects
	// Deprecated: Use ObjectConnector.Filter instead.
```

File: service/assets/object.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	// https://docs.go-atlassian.io/jira-assets/object#create-object
+	Create(ctx context.Context, workspaceID string, payload *models.ObjectPayloadScheme) (*models.ObjectScheme, *models.ResponseScheme, error)
+
+	// Relation returns the relation between Jira issues and Assets objects
+	//
+	// GET /jsm/assets/workspace/{workspaceId}/v1/objectconnectedtickets/{objectId}/tickets
+	//
+	// https://docs.go-atlassian.io/jira-assets/object#get-object-tickets
+	Relation(ctx context.Context, workspaceID, objectID string) (*models.TicketPageScheme, *models.ResponseScheme, error)
+}
```
Comment (Fank) [inline service/assets/object.go:67]:
```suggestion

	// Filter fetch Objects by AQL.
	//
	// POST /jsm/assets/workspace/{workspaceId}/v1/object/aql
	//
	// https://docs.go-atlassian.io/jira-assets/objects#filter-objects
	Filter(ctx context.Context, workspaceID string, payload *models.ObjectsAQLSearchParamsScheme) (*models.ObjectScheme, *models.ResponseScheme, error)
}
```

**Review by Fank — CHANGES_REQUESTED** [review]

File: service/assets/aql.go (hunk 2 above)
Comment (Fank) [reply service/assets/aql.go:16]: image

**Comment by ctreminiom** [comment]
> @ctreminiom it would be nice if you could provide some functions for checking errors, like in the case i had before:
> 
> ```go
> assetObject, err = jiraClient.Assets.Object.Get(ctx, jiraAssetsWorkspaceID, jiraAssetsID)
> if err != nil && !errors.Is(err, cloud.ErrNotFound) {
> 	return err
> }
> ```
> 
> Currently it would looks like
> 
> ```go
> assetObject, response, err := jiraAssetsClient.Object.Get(ctx, jiraAssetsWorkspaceID, jiraAssetsID)
> if err != nil && response.StatusCode != http.StatusNotFound {
> 	return err
> }
> ```
> 
> Its is not that bad, just wondering how you would handle such error cases? In this situation I want to check on an `Object.Get` if it exists if yes `Object.Update` if not `Object.Create`.

_________________

Hi @Fank, it's because there's a possibility the HTTP call should return an error `(timeout error)` and the _Call()_ method does not return the response struct due to the callback returns nothing.

But, I can put more logic on the HTTP implementation and manipulate the errors based on the HTTP status code so the error `cloud.ErrNotFound` can be used on the other modules.

**Comment by Fank** [comment]
@ctreminiom the reason why i wrote "deprecated" is because IDEs will display deprecation, which makes it easier for developer to identify it they are using deprecated functions.

Currently:
image

By using "Deprecated" instead of "Not supported":
image

**Comment by Fank** [comment]
missing docs for https://docs.go-atlassian.io/jira-assets/objects#filter-objects

**Comment by ChristianKniep** [comment]
Hey folks, digging deeper into how to populate assets it seems to me the cloud version prefers [External Imports](https://developer.atlassian.com/cloud/assets/imports-rest-api-guide/schema-and-mapping/).
It includes a process which creates a mapping (JSON) and uploads the data via rather simple http calls. The most work (it seems) is to create the types to generate the JSON and make the process robust. 
AFAIU this is not in scope for the `phase-1` of assets? I am going to give it a go anyway. I reckon contributing it to `go-atlassian` makes a lot of sense. Maybe in `phase-2`?  Fair assumptions?

**Comment by Fank** [comment]
I would recommend splitting it in another part.
This PR is part of https://github.com/ctreminiom/go-atlassian/issues/202 and i would recommend comment it here or add another PR and add your request to it.
To be honest i have never seen that External Import. Our need was realtime updates Assets, i think thats the reason.

**Review by ctreminiom — COMMENTED** [review]
image

**Comment by ctreminiom** [comment]
> missing docs for https://docs.go-atlassian.io/jira-assets/objects#filter-objects

Fixed 👍🏽

**Comment by ctreminiom** [comment]
> Looks like there is something odd with the atlassian docs, https://developer.atlassian.com/cloud/assets/rest/api-group-object/#api-object-aql-post i see you implemented `ObjectListResult` as `ObjectListScheme` but if you check the properties of it image it does not match not the example (which looks like the same from my tests) image
> 
> It looks correct for the AQL and IQL parts, but not for `/objects/aql`

My bad, I confused the `/object/navlist/aql `schema with ` /object/aql`, I've changed the schema on the commit [86656a3](https://github.com/ctreminiom/go-atlassian/pull/204/commits/86656a3598e740c865410b88f76709d18cf3a620)

**Comment by ctreminiom** [comment]
> Hey folks, digging deeper into how to populate assets it seems to me the cloud version prefers [External Imports](https://developer.atlassian.com/cloud/assets/imports-rest-api-guide/schema-and-mapping/). It includes a process which creates a mapping (JSON) and uploads the data via rather simple http calls. The most work (it seems) is to create the types to generate the JSON and make the process robust. AFAIU this is not in scope for the `phase-1` of assets? I am going to give it a go anyway. I reckon contributing it to `go-atlassian` makes a lot of sense. Maybe in `phase-2`? Fair assumptions?

Hey @ChristianKniep, the _phase-1_ doesn't have the External Imports feature, it only adds the ability to manipulate the Objects, AQL, Icons, Object Attributes, Object Schemas and Object Types.

image

I would suggest to you to create a new PR based on this one and map the External Import endpoints😄. 

image

@Fank there was some misunderstanding about the AQL service and the Object AQL services, but it's know fixed 👍🏽.

**Comment by ChristianKniep** [comment]
Hey there, using the asset Client I notice that the URL is hard coded. For testing it would be nicer if the host is part of the `New` function like with the others.
```
diff --git a/assets/api_client_impl.go b/assets/api_client_impl.go
index 1831ad7..5ebe45f 100644
--- a/assets/api_client_impl.go
+++ b/assets/api_client_impl.go
*snip import changes*
 const DefaultAssetsSite = "https://api.atlassian.com/"
 
-func New(httpClient common.HttpClient) (*Client, error) {
+func New(httpClient common.HttpClient, site string) (*Client, error) {
 
        if httpClient == nil {
                httpClient = http.DefaultClient
        }
 
-       u, err := url.Parse(DefaultAssetsSite)
+       if site == "" {
+               return nil, model.ErrNoSiteError
+       }
+
+       if !strings.HasSuffix(site, "/") {
+               site += "/"
+       }
+
+       u, err := url.Parse(site)
+
        if err != nil {
                return nil, err
        }
```

**Comment by ctreminiom** [comment]
@Fank @ChristianKniep do you have more feedback about this module?, just before the merging.

**Comment by ChristianKniep** [comment]
@ctreminiom I do not, looking forward to it being merged.. :)

**Comment by Fank** [comment]
@ctreminiom I would say it's good to go

**Comment by ctreminiom** [comment]
Merging now then....

**Pull request merged**